src/TesApi.Web/BatchPool.cs

```diff
@@ -0,0 +1,584 @@
+                            }
+                            else
+                            {
+                                await RemoveNodesAsync((IList<ComputeNode>)nodesToRemove, cancellationToken);
+                            }
+                        }
+                        break;
+
+                    case ScalingMode.RemovingFailedNodes:
+                        ResizeErrors.Clear();
+                        _resizeErrorsRetrieved = true;
+                        await _azureProxy.EnableBatchPoolAutoScaleAsync(Pool.PoolId, !IsDedicated, AutoScaleEvaluationInterval, AutoPoolFormula, cancellationToken);
+                        _autoScaleWaitTime = DateTime.UtcNow + AutoScaleEvaluationInterval;
+                        _scalingMode = _resizeStoppedReceived ? ScalingMode.WaitingForAutoScale : ScalingMode.SettingAutoScale;
+                        break;
+
+                    case ScalingMode.WaitingForAutoScale:
+                        _resizeStoppedReceived = false;
+                        if (DateTime.UtcNow > _autoScaleWaitTime)
+                        {
+                            _scalingMode = ScalingMode.SettingAutoScale;
+                        }
+                        break;
+
+                    case ScalingMode.SettingAutoScale:
+                        _scalingMode = ScalingMode.AutoScaleEnabled;
+                        break;
+                }
+            }
+
+            IAsyncEnumerable<ComputeNode> GetNodesToRemove(bool withState)
+                => _azureProxy.ListComputeNodesAsync(Pool.PoolId, new ODATADetailLevel(filterClause: @"state eq 'starttaskfailed' or state eq 'preempted' or state eq 'unusable'", selectClause: withState ? @"id,state,startTaskInfo" : @"id"));
+        }
+
+        private ValueTask ServicePoolRotateAsync(CancellationToken _1)
+        {
+            if (IsAvailable)
+            {
+                IsAvailable = Creation + _forcePoolRotationAge > DateTime.UtcNow;
+            }
+
+            return ValueTask.CompletedTask;
+        }
+
+        private async ValueTask ServicePoolRemovePoolIfEmptyAsync(CancellationToken cancellationToken)
+        {
+            if (!IsAvailable)
+            {
+                var (lowPriorityNodes, dedicatedNodes) = await _azureProxy.GetCurrentComputeNodesAsync(Pool.PoolId, cancellationToken);
+                if ((lowPriorityNodes is null || lowPriorityNodes == 0) && (dedicatedNodes is null || dedicatedNodes == 0) && !await GetTasksAsync().AnyAsync(cancellationToken))
+                {
+                    await _batchPools.DeletePoolAsync(this, cancellationToken);
+                    _ = _batchPools.RemovePoolFromList(this);
+                }
+            }
+        }
+    }
+
+    /// <content>
+    /// Implements the <see cref="IBatchPool"/> interface.
+    /// </content>
+    public sealed partial class BatchPool : IBatchPool
+    {
+        private sealed class LockObj : IDisposable
```

Comment:
I wrapped it to get two behaviors (timed lock vs infinite lock) while keeping the same static underlying synchronization object and then later removed the timed lock behavior, but never refactored this. Removing the wrapper since it's not needed.